python/lsst/daf/persistence/storageInterface.py

```diff
@@ -64,28 +65,69 @@ def registerFormatter(cls, formatable, formatter):
+            The formatter class that can be used by the StorageInterface instance to write the object to the
+            storage.
 
         Raises
         ------
         RuntimeError
-            For each object type and StorageInterface subclass a formatter should only be registered once. If
-            a second registration occurs a RuntimeError is raised.
+            For each object type and StorageInterface subclass the read and write formatters should only be
+            registered once. If a second registration occurs for either a RuntimeError is raised.
         """
-        classFormatters = StorageInterface.formatters.setdefault(cls, {})
+        cls.registerReadFormatter(formatable, readFormatter)
+        cls.registerWriteFormatter(formatable, writeFormatter)
+
+    @classmethod
+    def registerReadFormatter(cls, formatable, readFormatter):
+        """Similar to registerFormatters, but only for the readFormatter."""
+        cls._registerFormatter(formatable, readFormatter, 'read')
+
+    @classmethod
+    def registerWriteFormatter(cls, formatable, writeFormatter):
+        """Similar to registerFormatters, but only for the writeFormatter."""
+        cls._registerFormatter(formatable, writeFormatter, 'write')
+
+    @classmethod
+    def _registerFormatter(cls, formatable, formatter, rw):
+        """Implementation of the registerFormatter... funcitons."""
+        if rw == 'read':
+            classFormatters = StorageInterface.readFormatters.setdefault(cls, {})
+        elif rw == 'write':
+            classFormatters = StorageInterface.writeFormatters.setdefault(cls, {})
+        else:
+            raise RuntimeError("invalid value for rw: {}".format(rw))
+
         if formatable in classFormatters:
-            raise RuntimeError(("Registration of second formatter {} for formattable class {} in " +
-                                " storageInterface {}").format(formatter, formatable, cls))
+            raise RuntimeError(("Registration of second {} formatter {} for formattable class {} in " +
+                                " storageInterface {}").format(rw, formatter, formatable, cls))
         classFormatters[formatable] = formatter
 
     @classmethod
-    def _getFormatter(cls, objType):
-        """Search in the registered formatters for the formatter for obj.
+    def _getReadFormatter(cls, objType):
+        """Search in the registered formatters for the objType read formatter.
 
-        Will attempt to find formatters registered for the objec type, and then
-        for base classes of the object in resolution order.
+        Parameters
+        ----------
+        objType : class type
+            The type of class to find a formatter for.
+
+        Returns
+        -------
+        formatter callable
+            The formatter callable used to read the object from the storageInterface.
+        """
+        classFormatters = StorageInterface.readFormatters.get(cls, None)
```

Comment:
One minor issue with using `cls` as a key in a global map — if, for example, I decide to implement new Storage as subclass of PosixStorage (not sure if that is possible) then calling `self._getFormatter()` in a subclass will not find any formatters registered with PosixStorage?

Reply:
yeah, it's broken. I ran into it the other day. Needs fixing.